server/issue.go

```diff
@@ -410,456 +410,468 @@ func (s *Server) getPipelineApprovalPolicyForEnv(ctx context.Context, envID int)
-				EnvironmentID:    database.Instance.Environment.ID,
-				TaskList:         taskCreateList,
-				TaskIndexDAGList: taskIndexDAGList,
+				Name:          fmt.Sprintf("%s %s", database.Instance.Environment.Name, database.Name),
+				EnvironmentID: database.Instance.Environment.ID,
+				TaskList:      []api.TaskCreate{*taskCreate},
 			})
 		}
-		return create, nil
+	}
+	return create, nil
+}
 
+func (s *Server) getPipelineCreateForDatabaseSchemaUpdateGhost(ctx context.Context, issueCreate *api.IssueCreate) (*api.PipelineCreate, error) {
+	if !s.feature(api.FeatureGhost) {
+		return nil, echo.NewHTTPError(http.StatusForbidden, api.FeatureGhost.AccessErrorMessage())
+	}
+	c := api.UpdateSchemaGhostContext{}
+	if err := json.Unmarshal([]byte(issueCreate.CreateContext), &c); err != nil {
+		return nil, err
+	}
+	if !s.feature(api.FeatureTaskScheduleTime) {
+		for _, detail := range c.DetailList {
+			if detail.EarliestAllowedTs != 0 {
+				return nil, echo.NewHTTPError(http.StatusForbidden, api.FeatureTaskScheduleTime.AccessErrorMessage())
+			}
+		}
+	}
+
+	project, err := s.store.GetProjectByID(ctx, issueCreate.ProjectID)
+	if err != nil {
+		return nil, echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("failed to fetch project with ID %d", issueCreate.ProjectID)).SetInternal(err)
+	}
+	if project.TenantMode == api.TenantModeTenant {
+		return nil, echo.NewHTTPError(http.StatusBadRequest, "not implemented yet")
+	}
+
+	create := &api.PipelineCreate{}
+	create.Name = "Update database schema (gh-ost) pipeline"
+	schemaVersion := common.DefaultMigrationVersion()
+	for _, detail := range c.DetailList {
+		if detail.Statement == "" {
+			return nil, echo.NewHTTPError(http.StatusBadRequest, "failed to create issue, sql statement missing")
+		}
+
+		database, err := s.store.GetDatabase(ctx, &api.DatabaseFind{ID: &detail.DatabaseID})
+		if err != nil {
+			return nil, echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("failed to fetch database ID: %v", detail.DatabaseID)).SetInternal(err)
+		}
+		if database == nil {
+			return nil, echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("database ID not found: %d", detail.DatabaseID))
+		}
+
+		taskStatus, err := s.getPipelineApprovalPolicyForEnv(ctx, database.Instance.EnvironmentID)
+		if err != nil {
+			return nil, err
+		}
+
+		taskCreateList, taskIndexDAGList, err := createGhostTaskList(database, c.VCSPushEvent, detail, schemaVersion, taskStatus)
+		if err != nil {
+			return nil, err
+		}
+
+		create.StageList = append(create.StageList, api.StageCreate{
+			Name:             fmt.Sprintf("%s %s", database.Instance.Environment.Name, database.Name),
+			EnvironmentID:    database.Instance.Environment.ID,
+			TaskList:         taskCreateList,
+			TaskIndexDAGList: taskIndexDAGList,
+		})
+	}
+	return create, nil
+}
+
+func (s *Server) getPipelineCreate(ctx context.Context, issueCreate *api.IssueCreate) (*api.PipelineCreate, error) {
```

Comment:
nit: move this function above function getPipelineCreateForDatabaseCreate(). Typically, most important information first. The caller, which is more important, should come before the callee.

Reply:
SG.